File: app/src/main/java/com/main/climbingdiary/fragments/SettingsFragment.java
package com.main.climbingdiary.fragments;

import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.preference.Preference;
import android.preference.PreferenceFragment;
import android.preference.PreferenceManager;
import android.util.Log;

import com.main.climbingdiary.R;
import com.main.climbingdiary.activities.MainActivity;
import com.main.climbingdiary.activities.SettingsActivity;
import com.main.climbingdiary.common.AlertManager;
import com.main.climbingdiary.common.AppFileProvider;
import com.main.climbingdiary.common.preferences.AppPreferenceManager;
import com.main.climbingdiary.common.preferences.PreferenceKeys;
import com.main.climbingdiary.models.Alert;

import java.io.IOException;

import cn.pedant.SweetAlert.SweetAlertDialog;
import lombok.var;

import static android.app.Activity.RESULT_OK;
import static com.main.climbingdiary.common.preferences.PreferenceKeys.FILE_CHOOOSER_REQUEST_RESTORE_COPY;
import static com.main.climbingdiary.common.preferences.PreferenceKeys.FILE_CHOOOSER_REQUEST_SAFTY_COPY;

public class SettingsFragment extends PreferenceFragment {

    private Preference dbOutputPath;
    private static volatile SettingsFragment INSTANZ = null;

    public synchronized static SettingsFragment getInstance() {
        if (INSTANZ == null) {
            INSTANZ = new SettingsFragment();
        }
        return INSTANZ;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.settings);
        PreferenceManager.getDefaultSharedPreferences(MainActivity.getMainAppContext());

        dbOutputPath = getPreferenceManager().findPreference(PreferenceKeys.DB_OUTPUT_PATH);
        Preference shareDb = getPreferenceManager().findPreference(PreferenceKeys.SAFTY_COPY);
        Preference restoreDbPath = getPreferenceManager().findPreference(PreferenceKeys.RESTORE_COPY);
        /* ToDo
        Preference changeAreaName = getPreferenceManager().findPreference(PreferenceKeys.UPDATE_AREA);
        Preference changeSectorName = getPreferenceManager().findPreference(PreferenceKeys.UPDATE_SECTOR);*/

        dbOutputPath.setOnPreferenceClickListener(preference -> {
            this.openFolderChooser();
            return true;
        });
        assert restoreDbPath != null;
        restoreDbPath.setOnPreferenceClickListener(preference -> {
            this.openFileChooser();
            return true;
        });
        assert shareDb != null;
        shareDb.setOnPreferenceClickListener(preference -> {
            this.exportDb();
            return true;
        });
       /* ToDo
       assert changeAreaName != null;
        changeAreaName.setOnPreferenceClickListener(preference -> {
            this.changeAreaOrSectorName(changeAreaName.getKey());
            return true;
        });
        assert changeSectorName != null;
        changeSectorName.setOnPreferenceClickListener(preference -> {
            this.changeAreaOrSectorName(changeSectorName.getKey());
            return true;
        });*/
        this.initPrefs();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == FILE_CHOOOSER_REQUEST_SAFTY_COPY && resultCode == RESULT_OK) {
            Uri selectedfile = data.getData();
            assert selectedfile != null;
            AppPreferenceManager.setOutputPath(selectedfile.toString());
            dbOutputPath.setSummary(AppPreferenceManager.getOutputPath());
        }
        if (requestCode == FILE_CHOOOSER_REQUEST_RESTORE_COPY && resultCode == RESULT_OK) {
            Uri path = data.getData();
            this.restoreDB(path);
        }
    }

    private void initPrefs() {
        try {
            dbOutputPath.setSummary(AppPreferenceManager.getOutputPath());
        } catch (NullPointerException ignored) {
        }
    }

    private void openFolderChooser() {
        Intent i = new Intent(Intent.ACTION_OPEN_DOCUMENT_TREE);
        i.addCategory(Intent.CATEGORY_DEFAULT);
        startActivityForResult(Intent.createChooser(i, "Output-Pfad"), FILE_CHOOOSER_REQUEST_SAFTY_COPY);
    }

    private void openFileChooser() {
        Intent intent = new Intent(Intent.ACTION_GET_CONTENT);
        intent.setType("*/*");
        startActivityForResult(intent, FILE_CHOOOSER_REQUEST_RESTORE_COPY);
    }

    private void exportDb() {
        try {
            new AppFileProvider().exportDBtoPreferencePath();
            new AlertManager().setAlertWithoutContent(
                    this.getContext(),
                    Alert.builder().dialogType(SweetAlertDialog.SUCCESS_TYPE).title("Die Datenbank wurde exportiert !").build());
        } catch (IOException e) {
            Log.e("exportDb", e.getLocalizedMessage());
            new AlertManager().setAlertWithoutContent(
                    this.getContext(),
                    Alert.builder().dialogType(SweetAlertDialog.ERROR_TYPE)
                            .title(String.format("Der Export ist Schiefgelaufen %s", "\ud83d\ude13"))
                            .build());
        }
    }

    private void restoreDB(Uri path) {
        try {
            if (new AppFileProvider().restoreDBfromPreferencePath(path)) {
                new SweetAlertDialog(this.getContext(), SweetAlertDialog.SUCCESS_TYPE)
                        .setTitleText("Die Datenbank wurde wiederhergestellt !")
                        .setContentText("Die Anwendung wird neu gestartet.")
                        .setConfirmClickListener(sDialog -> {
                            var context = this.getContext();
                            PackageManager packageManager = context.getPackageManager();
                            Intent intent = packageManager.getLaunchIntentForPackage(context.getPackageName());
                            assert intent != null;
                            ComponentName componentName = intent.getComponent();
                            Intent mainIntent = Intent.makeRestartActivityTask(componentName);
                            context.startActivity(mainIntent);
                            Runtime.getRuntime().exit(0);
                        })
                        .show();
            }
        } catch (IOException e) {
            Log.d("restoreDb", e.getLocalizedMessage());
            new AlertManager().setAlertWithoutContent(
                    this.getContext(),
                    Alert.builder().dialogType(SweetAlertDialog.ERROR_TYPE)
                            .title(String.format("Der Restore ist Schiefgelaufen %s", "\ud83d\ude13"))
                            .build());
        }
    }

    private void changeAreaOrSectorName(String key) {
        SettingsActivity
                .getSettingsActivity()
                .getFragmentManager()
                .beginTransaction()
                .replace(android.R.id.content, new EditAreaFragment()).commit();

    }
}

File: app/src/main/java/com/main/climbingdiary/fragments/SettingsFragment.kt
package com.main.climbingdiary.fragments

import android.app.Activity.RESULT_OK
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import cn.pedant.SweetAlert.SweetAlertDialog
import com.main.climbingdiary.R
import com.main.climbingdiary.common.AlertFactory.getAlert
import com.main.climbingdiary.common.AppPermissions
import com.main.climbingdiary.common.LanguageManager
import com.main.climbingdiary.common.RessourceFinder
import com.main.climbingdiary.common.StringManager
import com.main.climbingdiary.common.preferences.AppPreferenceManager
import com.main.climbingdiary.common.preferences.AppPreferenceManager.getOutputPath
import com.main.climbingdiary.common.preferences.PreferenceKeys
import com.main.climbingdiary.controller.slider.TimeSliderFactory
import com.main.climbingdiary.models.Alert
import com.main.climbingdiary.models.TimeRange
import com.main.climbingdiary.provider.AppFileProvider
import java.io.File
import java.io.IOException

class SettingsFragment : PreferenceFragmentCompat() {

    private val dbOutputPath: Preference? by lazy {
        findPreference(RessourceFinder.getStringRessourceById(R.string.db_output_path))
    }
    private val fileChooserRequestCopy =
        PreferenceKeys.FILE_CHOOOSER_REQUEST_RESTORE_COPY
    private val fileChooserSafetyCopy = PreferenceKeys.FILE_CHOOOSER_REQUEST_SAFTY_COPY
    private val timeSliderView: ListPreference? by lazy {
        findPreference(RessourceFinder.getStringRessourceById(R.string.pref_time_slider))
    }
    private val languagePref:Preference? by lazy {
        findPreference(RessourceFinder.getStringRessourceById(R.string.language))
    }
    private val languageSet by lazy { AppPreferenceManager.getLanguage() }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.settings, rootKey)
        val exportDb: Preference? =
            findPreference(RessourceFinder.getStringRessourceById(R.string.safty_copy))

        val saftyCopy: Preference? =
            findPreference(RessourceFinder.getStringRessourceById(R.string.safty_restore))

        //set the correct flag
        languagePref.let {
            when(languageSet){
                "de"-> it!!.setIcon(R.drawable.en)
                "en"-> it!!.setIcon(R.drawable.de)
            }
        }

        dbOutputPath?.setOnPreferenceClickListener {
            openFolderChooser()
            true
        }
        saftyCopy?.setOnPreferenceClickListener {
            openFileChooser()
            true
        }
        exportDb?.setOnPreferenceClickListener {
            exportDb()
            true
        }

        timeSliderView?.setOnPreferenceChangeListener { _, newValue ->
            val range = TimeRange.translate(newValue as String)
            AppPreferenceManager.setTimeSliderView(range)
            timeSliderView!!.summary = newValue
            TimeSliderFactory.reloadSlider()
            true
        }

        languagePref?.setOnPreferenceClickListener {
            changeLocale()
            true
        }

        initPrefs()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == fileChooserSafetyCopy && resultCode == RESULT_OK) {
            val selectedfile = data?.data!!
            val paths: List<String> = selectedfile.path!!.split(":")
            val completePath = Environment.getExternalStorageDirectory().toString() +
                    if (paths.size > 1) File.separator + paths[1] else ""
            AppPreferenceManager.setOutputPath(completePath)
            dbOutputPath!!.summary = completePath

        }
        if (requestCode == fileChooserRequestCopy && resultCode == RESULT_OK) {
            val path = data?.data
            restoreDB(path)
        }
    }

    private fun initPrefs() {
        try {
            dbOutputPath!!.summary = getOutputPath()
            //set the default value
            timeSliderView!!.let {
                if (it.value == null) {
                    it.setValueIndex(0)
                }
            }
            timeSliderView!!.summary = TimeRange.translate(AppPreferenceManager.getTimeSliderView())
        } catch (ignored: NullPointerException) {
        }
    }

    private fun openFolderChooser() {
        val i = Intent(Intent.ACTION_OPEN_DOCUMENT_TREE)
        i.addCategory(Intent.CATEGORY_DEFAULT)
        i.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        startActivityForResult(
            Intent.createChooser(i, "Output-Pfad"),
            fileChooserSafetyCopy
        )
    }

    private fun openFileChooser() {
        val intent = Intent(Intent.ACTION_GET_CONTENT)
        intent.type = "*/*"
        startActivityForResult(intent, fileChooserRequestCopy)
    }

    private fun exportDb() {
        try {
            AppFileProvider().exportDBtoPreferencePath()
            this.context?.let {
                getAlert(
                    it,
                    Alert(
                        dialogType = SweetAlertDialog.SUCCESS_TYPE,
                        title = "Die Datenbank wurde exportiert !"
                    )
                ).show()
            }
        } catch (e: IOException) {
            this.context?.let { it ->
                val errorAlert = getAlert(
                    it,
                    Alert(
                        dialogType = SweetAlertDialog.ERROR_TYPE,
                        title = "Der Export ist Schiefgelaufen 😓"
                    )
                )
                errorAlert.setConfirmClickListener {
                    AppPermissions.checkFileManagementPermission(this.context)
                    it.cancel()
                }.show()
            }
        }
    }

    private fun restoreDB(path: Uri?) {
        try {
            if (path?.let { AppFileProvider().restoreDBfromPreferencePath(it) } == true) {
                val alert = getAlert(requireContext(),
                    Alert(title=getString(R.string.alert_db_restored_correctly),
                        message = getString(R.string.app_will_restart) ))
                    .setConfirmClickListener {
                        val context: Context? = this.context
                        val packageManager: PackageManager = context!!.packageManager
                        val intent =
                            packageManager.getLaunchIntentForPackage(context.packageName)!!
                        val componentName = intent.component
                        val mainIntent = Intent.makeRestartActivityTask(componentName)
                        context.startActivity(mainIntent)
                        Runtime.getRuntime().exit(0)
                    }
                    alert.show()
            }
        } catch (e: IOException) {
            Log.d("restoreDb", e.localizedMessage as String)
            this.context?.let {
                getAlert(
                    it,
                    Alert(
                        dialogType = SweetAlertDialog.ERROR_TYPE,
                        title = getString(R.string.alert_restore_db_failure)
                    )
                ).show()
            }
        }
    }

    private fun changeLocale(){
        LanguageManager(requireContext())
            .switchLanguage()
            .let {
                val alert = getAlert(requireContext(),
                    Alert(title=StringManager.getStringForId(R.string.alert_language_change),
                        message = StringManager.getStringForId(R.string.app_will_restart) ))
                    .setConfirmClickListener {
                        val context: Context? = context
                        val packageManager: PackageManager = context!!.packageManager
                        val intent =
                            packageManager.getLaunchIntentForPackage(context.packageName)!!
                        val componentName = intent.component
                        val mainIntent = Intent.makeRestartActivityTask(componentName)
                        context.startActivity(mainIntent)
                        Runtime.getRuntime().exit(0)
                    }
                alert.show()
            }
    }
}